### PR TITLE
Adding support for tree dynamic connectivity

### PR DESCRIPTION
**Pull request by Toptachamann**
### Current state
Currently, we have a union-find data structures, which solves the graph incremental connectivity problem. 
### Goal
The goal of this PR is to extend the libraries support of the dynamic graphs. This PR contains an implementation of the [Euler tour technique](https://en.wikipedia.org/wiki/Euler_tour_technique) to store trees dynamically and query the node connectivity in log(n) time. 
### Benchmarking: 
The benchmarks were done as follows:
* generate 2 trees using `BarabasiAlbertForestGenerator`
* take 5 random vertices from each tree
* Perform the connectivity requests for each pair of vertices (25 in total):
   * For Euler tour trees a) connect 2 vertices b) query connectivity c) disconnect vertices (connecting vertices are not equal to the query vertices so that the query is not biased)
   * For naive algorithm, add an edge to the graph and then run BFS
* Divide the resulting running time by 25 to get an average query time

Results:
euler
comp

### Conclusion
- As can be seen from the first log scale plot, the algorithm scales as a straight line => the complexity is logarithmic
- The performance improvement is high compared with the naive approach

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

**Comment by Toptachamann** [comment]
__Update:__ I've added the benchmark results to the PR description. Other libraries such as networkx, lemon or boost do not seem to have this algorithm, so there's no comparison with an existing implementation from another library

File: jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/TreeDynamicConnectivity.java [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+     */
+    public TreeDynamicConnectivity() {
+        minToTreeMap = new HashMap<>();
+        nodeMap = new HashMap<>();
+        singletonNodes = new HashMap<>();
+    }
+
+    /**
+     * Adds an {@code element} to this data structure. If the {@code element} has
+     * been added before, this method returns {@code false} and has node effect.
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/TreeDynamicConnectivity.java:79]: "node" -> "no"

File: jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/TreeDynamicConnectivity.java [hunk 2]
```diff
@@ -0,0 +1,629 @@
...
+import org.jgrapht.util.DoublyLinkedList;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static org.jgrapht.util.DoublyLinkedList.ListNode;
+import static org.jgrapht.util.AVLTree.TreeNode;
+
+/**
+ * Data structure for storing dynamic trees and querying node connectivity
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/alg/connectivity/TreeDynamicConnectivity.java:30]: I assume you are using the Euler tour data structure, so we should point to the proper reference which is: 

```
Monika Rauch Henzinger, Valerie King: Randomized dynamic graph algorithms with polylogarithmic time per operation. STOC 1995: 519-527
```

File: jgrapht-core/src/main/java/org/jgrapht/util/AVLTree.java [hunk 3]
```diff
@@ -0,0 +1,1091 @@
...
+     */
+    @Override
+    public String toString() {
+        StringBuilder builder = new StringBuilder();
+        for (Iterator<TreeNode<T>> i = nodeIterator(); i.hasNext(); ) {
+            TreeNode<T> node = i.next();
+            builder.append(node.toString()).append("\n");
+        }
+        return builder.toString();
+//        return "";
```
Comment (d-michail) [inline jgrapht-core/src/main/java/org/jgrapht/util/AVLTree.java:576]: Forgotten!

**Comment by d-michail** [comment]
Merged! Thanks!

**Pull request merged**